docs/src/main/sphinx/develop/supporting-merge.rst

```diff
@@ -0,0 +1,389 @@
+  operations.
+
+* ``finish``::
+
+    ``CompletableFuture<Collection<Slice>> finish()``
+
+  The Trino engine calls ``finish()`` when all the data has been processed by
+  a specific ``ConnectorMergeSink`` instance.  The connector returns a future
+  containing a collection of ``Slice``, representing connector-specific
+  information about the rows processed.  Usually this includes the row count,
+  and might include information like the files or partitions created or
+  changed.
+
+``ConnectorMetadata`` ``MERGE`` API
+-----------------------------------
+
+A connector implementing ``MERGE`` must implement these ``ConnectorMetadata``
+methods.
+
+* ``getRowChangeParadigm()``::
+
+    RowChangeParadigm getRowChangeParadigm(
+        ConnectorSession session,
+        ConnectorTableHandle tableHandle)
+
+  This method is called as the engine starts processing a ``MERGE`` statement.
+  The connector must return a ``RowChangeParadigm`` enum instance.  If the
+  connector does not support ``MERGE`` it should throw a ``NOT_SUPPORTED``
+  exception, meaning that ``SQL MERGE`` is not supported by the connector.
+
+* ``getMergeRowIdColumnHandle()``::
+
+    ColumnHandle getMergeRowIdColumnHandle(
+        ConnectorSession session,
+        ConnectorTableHandle tableHandle,
+        MergeDetails mergeDetails)
+
+  This method is called in the early stages of query planning for ``MERGE``
+  statements.  The ColumnHandle returned provides the ``rowId`` used by the
+  connector to identify rows to be merged, as well as any other fields of
+  the row that the connector needs to complete the ``MERGE`` operation.
+
+* ``getWriteRedistributionColumns()``::
+
+    List<ColumnHandle> getWriteRedistributionColumns(
+        ConnectorSession session,
+        ConnectorTableHandle table)
+
+  This method returns a list of ``ColumnHandles`` for table columns that
+  impact write redistribution, e.g., columns that impact partitioning or
+  bucketing.  By default, this method returns an empty list.
+
+* ``beginMerge()``::
+
+    ConnectorMergeTableHandle beginMerge(
+         ConnectorSession session,
+         ConnectorTableHandle tableHandle,
+         MergeDetails mergeDetails)
+
+  As the last step in creating the ``MERGE`` execution plan, the connector's
+  ``beginMerge()`` method is called, passing the ``session``, the
+  ``tableHandle`` and the ``MergeDetails`` object.
+
+  ``beginMerge()`` performs any orchestration needed in the connector to
+  start processing the ``MERGE``.  This orchestration varies from connector
+  to connector.  In the Hive ACID connector, for example, ``beginMerge()``
+  checks that the table is transactional and that all updated columns are
+  writable, and starts a Hive Metastore transaction.
+
+  ``beginMerge()`` returns a ``ConnectorMergeTableHandle`` with any added
+  information the connector needs when the handle is passed back to
+  ``finishMerge()`` and the split generation machinery.  For most
```

Comment:
is ConnectorMergeTableHandle passed to Split source, or is it `ConnectorMergeTableHandle.getTableHandle()`?

Reply:
As the text says, it's the `ConnectorMergeTableHandle`.  Perhaps I don't understand the question.

Reply:
among the modified files i didn't find the split-related things, so wondered how ConnectorMergeTableHandle gets passed to split generation machinery